### ISSUE TITLE
False positive on cyclic definition despite using forward reference in NamedTuples

### ISSUE THREAD

**Issue by lbenezriravin** [python/mypy#8671]
*Bug*

Environment:
mypy @ current master and python 3.7.4
```
$ pipenv graph
mypy==0.770+dev.2bbe22a97e472e8a325786c87417ac1bc21a7fd4
  - mypy-extensions [required: >=0.4.3,<0.5.0, installed: 0.4.3]
  - typed-ast [required: >=1.4.0,<1.5.0, installed: 1.4.1]
  - typing-extensions [required: >=3.7.4, installed: 3.7.4.2]
```

*Steps to reproduce*
Run mypy against the following code:
```
pipenv run mypy foo.py
```

```
from dataclasses import dataclass
from typing import NamedTuple


class Foo(NamedTuple):
    bar: "Foo"


foo = Foo(Foo(None))


class Foo2:
    def __init__(self, bar: "Foo2") -> None:
        pass


foo2 = Foo2(Foo2(None))


@dataclass
class Foo3:
    bar: "Foo3"


foo3 = Foo3(Foo3(None))
```

*Observed behavior*
```
foo.py:6: error: Cannot resolve name "Foo" (possible cyclic definition)
```

*Expected behavior*
No errors in the type definitions. No runtime errors are observed when actually running the code, as would normally be if `Foo` were truly unresolvable in that context. The other two examples (Foo2 and Foo3) behave as expected, and raise warnings about incompatible types where `None` is passed as an argument.

Thank you all for your work on this project!

**Comment by henri-v** [comment]
I don't think it's wrong to give an error in this case. Since Foo is immutable, the value of bar must be known at the time an instance is created. This means that to create a Foo instance you must first create a Foo instance, and therefore it is never possible to create a Foo instance in code that is type-correct.

However, consider this:
```py
from typing import NamedTuple, Optional


class Foo(NamedTuple):
    bar: Optional["Foo"]


foo = Foo(Foo(None))
```
Here I would not expect any errors. mypy correctly accepts `Foo(Foo(None))`, but here too, it fails with a possible cyclic definition at the definition of member bar:
```
test.py:5: error: Cannot resolve name "Foo" (possible cyclic definition)
```

**Comment by JukkaL** [comment]
If/when we support recursive types, the original example will also likely work (see #731). Otherwise we won't support recursive named tuples.

**Comment by lbenezriravin** [comment]
Ah, make sense to see that error if recursive types aren't supported. Thanks for the response, despite the duplicate issue.
